### DISCUSSION TITLE
OPC UA Query Multiple Nodes

### DISCUSSION THREAD

**Discussion by Praveenstein** [apache/streampipes#888] [Q&A]
How do i use one opc ua adapter to query multiple node ids at once??

**Comment by tenthe** [comment]
Currently, you can only query for one 'parent' node. 
The system then queries all the child nodes and displayes them in a tree. When you check the box the node will be be added to the resulting event. 
Is this approach sufficient for your use case or do you need a way to enter multiple different node ids manually?

**Reply by Praveenstein** [reply]
Thanks for the reply,I have three parameters with the node id given below

  - /Channel/ChannelDiagnose/cycleTime
  - /Channel/ChannelDiagnose/operatingTime
  - /Nck/ChannelDiagnose/setupTime
Now i can query the common parent  "/Channel/ChannelDiagnose" for the first two parameters and then check only cycleTime and operatingTime, but for the last parameter it has a different parent node id, and when i try to use root at the parent, it sometimes shows the following error 
![e4](https://user-images.githubusercontent.com/45847217/209318323-a59c25c9-a803-47f6-a943-94f6827703fd.png)

And it will be better if i can manually select multiple node ids, since i want to upload them as a single row in a database

**Reply by tenthe** [reply]
Ah ok, thanks for the reply.
That is a good point, we should open a discussion for a feature request.
I am not quite sure yet how the best solution would look like.
Maybe we can have a search field that can be used to browse the nodes and then a 'results' list that contains all selected nodes.
But I'm also not sure how we can do this generically.

Do you have an idea what would be best from a users perspevtive?

For now you can create two adapters and then use the processor **Merge Two Streams** in a pipeline to merge the data of the two adapters before storing them to the database.

**Reply by bossenti** [reply]
I have created a feature request [#890](https://github.com/apache/streampipes/discussions/890).
@Praveenstein please feel free to add your ideas and requirements there

**Reply by Praveenstein** [reply]
Thanks

**Reply by Praveenstein** [reply]
Just out of curiosity, let's say I have two adaptors connecting to same OPC server, but are querying two different set of node IDs, does the Apache streampipes Internally create two connections?? Or have one and share it among adaptors?? Because some OPC servers doesn't allow more connections at a time

**Reply by tenthe** [reply]
Currently, each adapter has its own connection. We would need some kind of connection sharing between the adapters.
Is this a common limitation of OPC servers or rather rarely the case?

**Reply by Praveenstein** [reply]
So I basically connect to cnc machines, currently to seimens controllers (through OPC ua), and they have configured it in such a way that after more than two connections it won't accept any more connections or will drop some existing connections.

**Reply by Praveenstein** [reply]
It all depends on the OPC server configuration and the hardware it runs on .... If i create my own server using official c# stack in a good hardware i will configure it to accept more connections, but i guess since cnc machine controllers don't have much hardware they're dropping connections

**Reply by tenthe** [reply]
Currently each adapter runs independently.
This means that there are no dependencies between them.

Do you think it would be good that multiple adapters share an OPC UA connection? 
Can this have any disadvantages? Because we do not know in particular how the users will configure the adapter (e.g. different settings).

**Reply by Praveenstein** [reply]
So basically any opc ua server will have only these configurations, ip address, port number, user name, password, (May be some security related settings). So one way to do is use lru (least recently used ) strategy. In python we have a built in decorator function lru_cache, what it does, for a given set of input parameters values, it will give one set of results (only the first time the results will be computed, after that it will return from the cache) here is a sample code for connecting to a opc server in python using this method:

from functools import lru_cache
import opcua

@lru_cache(maxsize=None)
def get_connection(host, port, user, password):
    return opcua.Client(host, port, user, password)

conn1 = get_connection('localhost', 4840, 'user1', 'password1')
conn2 = get_connection('localhost', 4840, 'user2', 'password2')
conn3 = get_connection('localhost', 4840, 'user1', 'password1')

# conn1 and conn3 will be the same connection object,
# because they have the same host, port, user, and password.
# conn2 will be a different connection object.


I guess apache streampipes in built in Java, i don't much proficiency in Java, but i tried using chatGpt, you can check it out:

import org.eclipse.milo.opcua.sdk.client.OpcUaClient;
import org.eclipse.milo.opcua.stack.client.UaTcpStackClient;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

@Service
public class OPCUAService {

    @Cacheable(value = "opcua_connections", key = "#host + ':' + #port + ':' + #user + ':' + #password")
    public OpcUaClient getConnection(String host, int port, String user, String password) throws Exception {
        return UaTcpStackClient.create(host, port)
            .setApplicationName(user)
            .setApplicationUri("urn:localhost:ua:client")
            .setCertificate(null)
            .setKeyPair(null)
            .setEndpointUrl(String.format("opc.tcp://%s:%d", host, port))
            .build();
    }
}

Hope this helps

**Reply by Praveenstein** [reply]
So yes it's better to share an adaptor (connection) for a given set of configuration

**Reply by bossenti** [reply]
@dominikriemer @tenthe sharing connections via caching seems to be a low hanging fruit for me. Do you agree? If so, we could create an issue for it and work on it soon.
